File: src/engine/schema.js
"use strict";

const fs = require("fs");
const path = require("path");
const Ajv = require("ajv");
const addFormats = require("ajv-formats");
const {
  DEFAULT_HISTORY_LIMIT,
  DEFAULT_HISTORY_RETENTION_DAYS,
  DEFAULT_FAILURE_POLICY,
  DEFAULT_MISFIRE_CAP,
  DEFAULT_MISFIRE_POLICY,
  DEFAULT_RETRIES,
  DEFAULT_RETRY_BACKOFF_SECONDS,
  DEFAULT_TIMEOUT_SECONDS,
  DEFAULT_TIMEZONE,
  buildHttpSpec,
  buildScriptSpec,
} = require("./models");

class ValidationError extends Error {}

const SCHEMA_PATH = path.resolve(__dirname, "../../schema/job.schema.json");
let validateFn = null;

function loadValidator() {
  if (!validateFn) {
    const ajv = new Ajv({ allErrors: true, strict: false });
    addFormats(ajv);
    validateFn = ajv.compile(JSON.parse(fs.readFileSync(SCHEMA_PATH, "utf8")));
  }
  return validateFn;
}

function validateMerged(merged, filePath) {
  const validate = loadValidator();
  if (validate(merged)) return;
  const messages = validate.errors.map((err) => {
    const loc = err.instancePath ? err.instancePath.slice(1).replaceAll("/", "/") : "<root>";
    return `${filePath}: ${err.message} (at ${loc})`;
  });
  throw new ValidationError(messages.join("\n"));
}

function buildJob(merged, filePath) {
  const notify = merged.notify || {};
  const topTimeout = merged.timeout_seconds || DEFAULT_TIMEOUT_SECONDS;
  const job = {
    id: merged.id,
    schedule: merged.schedule,
    type: merged.type,
    file_path: String(filePath),
    name: merged.name || merged.id,
    description: merged.description || "",
    timezone: merged.timezone || DEFAULT_TIMEZONE,
    enabled: merged.enabled ?? true,
    retries: merged.retries ?? DEFAULT_RETRIES,
    retry_backoff_seconds: merged.retry_backoff_seconds ?? DEFAULT_RETRY_BACKOFF_SECONDS,
    timeout_seconds: topTimeout,
    misfire_policy: merged.misfire_policy || DEFAULT_MISFIRE_POLICY,
    misfire_cap: merged.misfire_cap || DEFAULT_MISFIRE_CAP,
    history_limit: merged.history_limit || DEFAULT_HISTORY_LIMIT,
    history_retention_days: merged.history_retention_days || DEFAULT_HISTORY_RETENTION_DAYS,
    failure_policy: { ...DEFAULT_FAILURE_POLICY, ...(merged.failure_policy || {}) },
    notify_on_failure: notify.on_failure ?? true,
    notify_on_recovery: notify.on_recovery ?? true,
    http: null,
    script: null,
  };
  if (merged.type === "http") {
    job.http = buildHttpSpec(merged.http, topTimeout);
  } else {
    job.script = buildScriptSpec(merged.script, topTimeout);
  }
  return job;
}

function checkDuplicateIds(jobs) {
  const seen = new Map();
  const dupes = [];
  for (const job of jobs) {
    if (seen.has(job.id)) {
      dupes.push(`duplicate job id '${job.id}': ${seen.get(job.id)} and ${job.file_path}`);
    } else {
      seen.set(job.id, job.file_path);
    }
  }
  if (dupes.length) throw new ValidationError(dupes.join("\n"));
}

module.exports = { ValidationError, validateMerged, buildJob, checkDuplicateIds };

File: engine/schema.py
from __future__ import annotations

import json
from pathlib import Path

import jsonschema

from .models import (
    DEFAULT_HISTORY_LIMIT,
    DEFAULT_MISFIRE_CAP,
    DEFAULT_MISFIRE_POLICY,
    DEFAULT_RETRIES,
    DEFAULT_RETRY_BACKOFF_SECONDS,
    DEFAULT_TIMEOUT_SECONDS,
    DEFAULT_TIMEZONE,
    HttpSpec,
    Job,
    ScriptSpec,
)

SCHEMA_PATH = Path(__file__).resolve().parent.parent / "schema" / "job.schema.json"
_schema_cache: dict | None = None


class ValidationError(Exception):
    """Raised for schema violations or duplicate job ids. May bundle multiple messages."""


def _load_schema() -> dict:
    global _schema_cache
    if _schema_cache is None:
        with open(SCHEMA_PATH) as f:
            _schema_cache = json.load(f)
    return _schema_cache


def validate_merged(merged: dict, file_path: Path) -> None:
    validator = jsonschema.Draft7Validator(_load_schema())
    errors = sorted(validator.iter_errors(merged), key=lambda e: list(e.path))
    if errors:
        messages = [
            f"{file_path}: {e.message} (at {'/'.join(str(p) for p in e.path) or '<root>'})"
            for e in errors
        ]
        raise ValidationError("\n".join(messages))


def build_job(merged: dict, file_path: Path) -> Job:
    notify = merged.get("notify", {})
    top_timeout = merged.get("timeout_seconds", DEFAULT_TIMEOUT_SECONDS)

    http_spec = None
    script_spec = None
    if merged["type"] == "http":
        h = merged["http"]
        http_spec = HttpSpec(
            url=h["url"],
            method=h.get("method", "GET"),
            headers=h.get("headers", {}),
            body=h.get("body", ""),
            expected_status=h.get("expected_status", list(range(200, 300))),
            validate_contains=h.get("validate_contains"),
            timeout_seconds=h.get("timeout_seconds", top_timeout),
        )
    else:
        s = merged["script"]
        script_spec = ScriptSpec(
            path=s["path"],
            args=s.get("args", []),
            interpreter=s.get("interpreter"),
            timeout_seconds=s.get("timeout_seconds", top_timeout),
        )

    return Job(
        id=merged["id"],
        schedule=merged["schedule"],
        type=merged["type"],
        file_path=str(file_path),
        name=merged.get("name", merged["id"]),
        description=merged.get("description", ""),
        timezone=merged.get("timezone", DEFAULT_TIMEZONE),
        enabled=merged.get("enabled", True),
        retries=merged.get("retries", DEFAULT_RETRIES),
        retry_backoff_seconds=merged.get("retry_backoff_seconds", DEFAULT_RETRY_BACKOFF_SECONDS),
        timeout_seconds=top_timeout,
        misfire_policy=merged.get("misfire_policy", DEFAULT_MISFIRE_POLICY),
        misfire_cap=merged.get("misfire_cap", DEFAULT_MISFIRE_CAP),
        history_limit=merged.get("history_limit", DEFAULT_HISTORY_LIMIT),
        notify_on_failure=notify.get("on_failure", True),
        notify_on_recovery=notify.get("on_recovery", True),
        http=http_spec,
        script=script_spec,
    )


def check_duplicate_ids(jobs: list[Job]) -> None:
    seen: dict[str, str] = {}
    dupes: list[str] = []
    for job in jobs:
        if job.id in seen:
            dupes.append(f"duplicate job id '{job.id}': {seen[job.id]} and {job.file_path}")
        else:
            seen[job.id] = job.file_path
    if dupes:
        raise ValidationError("\n".join(dupes))
